interp: for loop with nil Post panics

Example script to reproduce this (in gosh, or a minimal parse/run go program):

```shell-script
for((i=0;i<=30;)) do echo "$i"; i=$((i+10)); done
```

The typical `++i` is missing here, and it seems to be valid in bash, but the parser simply puts `nil` as `Post`, and the interpreter does not expect it, and panics. While it works fine in GNU bash 5.0.17 for me.

```
panic: unexpected arithm expr: <nil>

goroutine 1 [running]:
mvdan.cc/sh/v3/expand.Arithm(0xc0000a0dc0, 0x0, 0x0, 0x1, 0x1, 0xc0000ac900)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/expand/arith.go:97 +0x89a
mvdan.cc/sh/v3/interp.(*Runner).arithm(0xc0000ac900, 0x0, 0x0, 0xc0002ac000)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/interp/runner.go:133 +0x46
mvdan.cc/sh/v3/interp.(*Runner).cmd(0xc0000ac900, 0x140f000, 0xc00012e000, 0x140adc0, 0xc000282e40)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/interp/runner.go:425 +0x158
mvdan.cc/sh/v3/interp.(*Runner).stmtSync(0xc0000ac900, 0x140f000, 0xc00012e000, 0xc0002a4000)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/interp/runner.go:275 +0x312
mvdan.cc/sh/v3/interp.(*Runner).stmt(0xc0000ac900, 0x140f000, 0xc00012e000, 0xc0002a4000)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/interp/runner.go:256 +0x1e9
mvdan.cc/sh/v3/interp.(*Runner).stmts(0xc0000ac900, 0x140f000, 0xc00012e000, 0xc0002ac020, 0x1, 0x4)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/interp/runner.go:586 +0x56
mvdan.cc/sh/v3/interp.(*Runner).Run(0xc0000ac900, 0x140f000, 0xc00012e000, 0x14013c0, 0xc0000fe880, 0x0, 0x0)
        /home/proto/go/pkg/mod/mvdan.cc/sh/v3@v3.1.2/interp/api.go:523 +0x2ec
```

Version v3.1.2

Let me know if you need more details, or if you can give me some pointers on how to go about making a PR to this project. This library is awesome so far btw! I am building a niche stateful shell with it for debugging peer to peer applications interactively :)

--- a/interp/interp_test.go
+++ b/interp/interp_test.go
@@ -629,6 +629,23 @@ var runTests = []runTest{
 		"for ((i=0; i<3; i++)); do echo $i; done",
 		"0\n1\n2\n",
 	},
+	// for, with missing Init, Cond, Post
+	{
+		"i=0; for ((; i<3; i++)); do echo $i; done",
+		"0\n1\n2\n",
+	},
+	{
+		"for ((i=0;; i++)); do if [ $i -ge 3 ]; then break; fi; echo $i; done",
+		"0\n1\n2\n",
+	},
+	{
+		"for ((i=0; i<3;)); do echo $i; i=$((i+1)); done",
+		"0\n1\n2\n",
+	},
+	{
+		"i=0; for ((;;)); do if [ $i -ge 3 ]; then break; fi; echo $i; i=$((i+1)); done",
+		"0\n1\n2\n",
+	},
 	// TODO: uncomment once expandEnv.Set starts returning errors
 	// {
 	// 	"readonly i; for ((i=0; i<3; i++)); do echo $i; done",

--- a/interp/runner.go
+++ b/interp/runner.go
@@ -430,12 +430,16 @@ func (r *Runner) cmd(ctx context.Context, cm syntax.Command) {
 				}
 			}
 		case *syntax.CStyleLoop:
-			r.arithm(y.Init)
-			for r.arithm(y.Cond) != 0 {
+			if y.Init != nil {
+				r.arithm(y.Init)
+			}
+			for y.Cond == nil || r.arithm(y.Cond) != 0 {
 				if r.exit != 0 || r.loopStmtsBroken(ctx, x.Do) {
 					break
 				}
-				r.arithm(y.Post)
+				if y.Post != nil {
+					r.arithm(y.Post)
+				}
 			}
 		}
 	case *syntax.FuncDecl:

--- a/syntax/nodes.go
+++ b/syntax/nodes.go
@@ -372,7 +372,8 @@ func (w *WordIter) End() Pos {
 //
 // This node will only appear with LangBash.
 type CStyleLoop struct {
-	Lparen, Rparen   Pos
+	Lparen, Rparen Pos
+	// Init, Cond, Post can each be nil, if the for loop construct omits it.
 	Init, Cond, Post ArithmExpr
 }
 

